Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -884,6 +884,34 @@ public struct ByteBuffer {
         }
         return true
     }
+    
+    /// The `ByteBuffer` will successfully be shrunk if the requested capacity is less than the current capacity, 
+    /// and the requested capacity is more than the number of readable bytes in the buffer.
+    /// If either condition is not true, the buffer will not be shrunk.
+    ///
+    /// - Parameter desiredCapacity: The desired capacity for the buffers capacity to be shrunken to
+    /// - Returns: Bool indicating whether the buffer capacity has been shrunk to the desiredCapacity.
+    @inlinable
+    @discardableResult public mutating func shrinkBufferCapacity(to desiredCapacity: Int) -> Bool {
+        let desiredCapacity = ByteBuffer.addPaddingTo(desiredCapacity)
+        guard desiredCapacity < capacity, desiredCapacity > readableBytes else {
+            return false
+        }
+        
+        let shrunkenCapacity = _toCapacity(desiredCapacity)
+        self._storage = self._storage.reallocSlice(self._slice.lowerBound..<self._slice.lowerBound + shrunkenCapacity, capacity: shrunkenCapacity)
```

Comment:
I think this reads potentially unwritten bytes which is UB.

Let's say we have an old ByteBuffer with `capacity: 8` in this state

```
slice.lowerBound (0)
| readerIndex (2)
| | writerIndex (4)
| | |   slice.upperBound (8, one past the end of the slice)
| | |   |
v v v   v
uuWWuuuu
```

If we now call `shrinkBufferCapacity(to: 4)` which will work, it's more than `self.readableBytes` but less than `self.capacity`, then we'll allocate a new storage of size 4 and copy `uuWW` in using `memcpy` in `reallocSlice`, that's UB.

where:

- `W`: is a byte written to (and readable)
- `u`: is a byte never written

We could create such a buffer as

```
var buf = ByteBuffer(capacity: 8)
buf.setWriterIndex(to: 2)
buf.setReaderIndex(to: 2)
buf.writeString("WW")
buf.shrinkBufferCapacity(to: 4)
XCTAssertEqual("WW", String(buffer: buf))
```

if this is run through Valgrind, I would expect valgrind to complain that we read bytes that were never written.

Reply:
@ali-ahsan-ali regardless of the potential UB, would you mind adding the above test case? The existing tests don't test anything with `readerIndex != 0`

Also worth testing:

```
var buf = ByteBuffer(capacity: 16384)
buf.setWriterIndex(to: 16000)
buf.setReaderIndex(to: 16000)
buf.writeString("WW")
buf.shrinkBufferCapacity(to: 4)
XCTAssertEqual("WW", String(buffer: buf))
```

Reply:
Why aren't we using `_copyStorageAndRebase(capacity: _Capacity, resetIndices: Bool = false)` actually?

Reply:
Added tests and using _copyStorageAndRebase with resettingIndices == true to avoid copying unwritten bytes. LMK if that addresses this comment

Reply:
@weissi Please take a look when you find the time.